arcs/src/systems/spatial_relation.rs

```diff
@@ -0,0 +1,204 @@
+            println!("Event is {:#?}", event);
+            match *event {
+                ComponentEvent::Inserted(id) => {
+                    self.to_insert.add(id);
+                },
+                ComponentEvent::Modified(id) => {
+                    self.to_update.add(id);
+                },
+                ComponentEvent::Removed(id) => {
+                    space.remove_by_id(id);
+                },
+            }
+        }
+
+        for (ent, drawing_object, _) in
+            (&entities, &drawing_objects, &self.to_insert).join()
+        {
+            let bb = drawing_object.geometry.bounding_box();
+            space.insert(SpatialEntity::new(bb, ent));
+        }
+
+        for (ent, drawing_object, _) in
+            (&entities, &drawing_objects, &self.to_update).join()
+        {
+            let bb = drawing_object.geometry.bounding_box();
+            space.modify(SpatialEntity::new(bb, ent));
+        }
+    }
+
+    fn setup(&mut self, world: &mut World) {
+        <Self::SystemData as shred::DynamicSystemData>::setup(
+            &self.accessor(),
+            world,
+        );
+
+        let drawing_storage = world.read_storage::<DrawingObject>();
+        let mut space = world.write_resource::<Space>();
+
+        space.clear();
+
+        for entity in world.entities().join() {
+            if let Some(drawing) = drawing_storage.get(entity) {
+                space.insert(SpatialEntity::new(drawing.geometry.bounding_box(), entity));
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        primitives::{Line},
+        components::{register, Layer, Name, DrawingObject, Geometry, LineStyle, Dimension, Space},
+        Vector,
+        systems::SpatialRelation,
+    };
+    use specs::prelude::*;
+    use piet::Color;
+
+    #[test]
+    fn setup_creates_all_outstanding_spatial_entities() {
+        let mut world = World::new();
+
+        // make sure we register all components
+        register(&mut world);
+    
+        let layer = Layer::create(
+            world.create_entity(),
+            Name::new("default"),
+            Layer {
+                z_level: 0,
+                visible: true,
+            },
+        );
+    
+        // Add a line to our world
+        let line = Line::new(Vector::new(2.0, 1.0), Vector::new(5.0, -1.0));
+        let _first = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Setup our spatial system
+        let mut system = SpatialRelation::new(&world);
+        System::setup(&mut system, &mut world);
+
+        let space = world.read_resource::<Space>();
+        assert_eq!(space.len(), 1);
```

Comment:
Can we also double-check the item in `space` is the `_first` entity?

Reply:
Done!